wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsUrl.java

```diff
@@ -100,14 +102,19 @@ public AzureDevOpsUrl withBranch(String branch) {
 
   @Override
   public String getProviderUrl() {
-    return "https://" + hostName;
+    return isNullOrEmpty(serverUrl) ? "https://" + hostName : serverUrl;
   }
 
   protected AzureDevOpsUrl withDevfileFilenames(List<String> devfileFilenames) {
     this.devfileFilenames.addAll(devfileFilenames);
     return this;
   }
 
+  public AzureDevOpsUrl withServerUrl(String serverUrl) {
+    this.serverUrl = serverUrl;
```

Comment:
When use set serverUrl, then hostname is null, which is used later in getRepositoryLocation

Reply:
The `hostName` does not rely on `serverUrl`, we set it independently:
https://github.com/eclipse-che/che-server/blob/4c865ce64cf46e331d224c5c12b6f7a9e09c0bd2/wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsURLParser.java#L198-L200